R/test-server.R

```diff
@@ -110,39 +110,86 @@ testModule <- function(module, expr, ...) {
   })
 }
 
+#' @noRd
+testCallModule <- function(module, id, session) {
+  # TODO alan Figure out what to do with id here, necessary for nested usage
+  body(module) <- rlang::expr({
+    session$env <- base::environment()
+    !!!body(module)
+  })
+
+  session$setReturned(do.call(module, list(
+    input = session$input,
+    output = session$output,
+    session = session
+  )))
+}
+
+# Create a "data mask" suitable for passing to rlang::eval_tidy. Bindings in
+# `env` and bindings in the parent of `env` are merged into a single named list.
+# Bindings in `env` take precedence over bindings in the parent of `env`.
+#' @noRd
+makeMask <- function(env) {
+  stopifnot(length(rlang::env_parents(env)) > 1)
+  stopifnot(all(c("input", "output", "session") %in% ls(env)))
+  child <- as.list(env)
+  parent <- as.list(rlang::env_parent(env))
+  parent_only <- setdiff(names(parent), names(child))
+  append(child, parent[parent_only])
+}
+
+#' @noRd
+isModuleServer <- function(x) {
+  is.function(x) && names(formals(x)) == "id"
+}
+
 #' Test an app's server-side logic
 #' @param appDir The directory root of the Shiny application. If `NULL`, this function
 #'   will work up the directory hierarchy --- starting with the current directory ---
 #'   looking for a directory that contains an `app.R` or `server.R` file.
 #' @rdname testModule
 #' @export
-testServer <- function(expr, appDir=NULL) {
-  if (is.null(appDir)){
-    appDir <- findApp()
-  }
-
-  app <- shinyAppDir(appDir)
-  message("Testing application found in: ", appDir)
-  server <- app$serverFuncSource()
+testServer <- function(app, expr, ...) {
 
-  origwd <- getwd()
-  setwd(appDir)
-  on.exit({ setwd(origwd) }, add=TRUE)
+  session <- MockShinySession$new()
+  on.exit(if (!session$isClosed()) session$close())
 
-  # Add `session` argument if not present
-  fn_formals <- formals(server)
-  if (! "session" %in% names(fn_formals)) {
-    fn_formals$session <- bquote()
-    formals(server) <- fn_formals
+  if (is.character(app) && length(app) == 1) {
+    server <- shinyAppDir(app)$serverFuncSource()
+    message("Testing application found in: ", app)
+    fn_formals <- formals(server)
+    if (! "session" %in% names(fn_formals)) {
```

Comment:
Instead of injecting `session`, we should just error and tell people that if they want to test, they need to have `session`. Also include that in docs.